Fixes #15750 - strong parameters (v1.13+)

lzap opened the pull request:
Do not merge before https://github.com/theforeman/foreman/pull/3659

Also prior merging this try to merge as many other PRs as possible to minimize
cherry-picking for 6.0 plugin release (which is 1.12 compatible).

This refactoring revealed some leaks in few functional tests - I was not
expecting return codes or result contents correctly and it was silently
failing.

lzap committed:
Fixes #15750 - strong parameters (v1.13+)

lzap commented on app/controllers/concerns/foreman/controller/parameters/discovered_host.rb:
```diff
@@ -0,0 +1,24 @@
+
+  def discovered_host_search_params
+    params.permit(:search, :order, :page)
+  end
```
These are not mass-assigned, I will remove the helper here and in `discovery_rule` as well.

domcleal commented on lib/foreman_discovery/engine.rb:
```diff
@@ -139,6 +139,11 @@ class Engine < ::Rails::Engine
+        # strong parameters
+        parameter_filter Subnet, :discovery_id
+        parameter_filter Host::Managed, :discovery_rule_id
+        parameter_filter Hostgroup, :discovery_rules, :discovery_rule_ids, :discovery_rule_names, :type
```
Should be `:type, :discovery_rules => [], :discovery_rule_ids => [], :discovery_rule_names => []` to permit an array of IDs/names.

lzap replied:
Thanks for review @domcleal !

lzap commented:
I have changed the parameter_filter and removed non-mass-assigned search params. Tests will fail - the core part need to be merged first.

lzap commented:
Added also `host.build` test in auto UI provisioning functional tests as described in https://github.com/theforeman/foreman_discovery/pull/289 (could not add them there because these tests were broken - asserts were incorrect).

domcleal commented on app/controllers/concerns/foreman/controller/parameters/discovered_host.rb:
```diff
@@ -0,0 +1,20 @@
+
+  class_methods do
+    def discovered_host_params_filter
+      Foreman::ParameterFilter.new(::Host::Discovered, :top_level_hash => 'host').tap do |filter|
```
:top_level_hash isn't accepted here any more, this second arg should be able to be removed.  (And it should now work for your API controller.)

lzap replied:
> :top_level_hash isn't accepted here any more, this second arg should be able to be removed.  (And it should now work for your API controller.)

Thanks for headsup, I will fix that and test your PR tomorrow.

## 

Later,
 Lukas #lzap Zapletal

lzap commented:
I rebased and removed the unnecessary param. I have some test failures locally but I believe this must be some migration issue. Pushing to see how Jenkins will like that.

domcleal commented on app/controllers/discovered_hosts_controller.rb:
```diff
@@ -50,7 +51,7 @@ def destroy
   def edit
     @host = ::ForemanDiscovery::HostConverter.to_managed(@host, true, false) unless @host.nil?
-    @host.attributes = @host.apply_inherited_attributes(params[:host]) unless params[:host].empty?
+    @host.attributes = @host.apply_inherited_attributes(discovered_host_params) unless discovered_host_params.empty?
```
You may need to add a `:host` argument to discovered_host_params and pass that through as the second arg to `filter_params` to specify the top level hash, as it won't be the standard `:discovered_host`.  Note that the API controller _does_ use :discovered_host, so the arg should default to `nil` and not be set in that controller.

Edit: i.e. `self.class.discovered_host_params_filter.filter_params(params, parameter_filter_context, :host)` from here, but `self.class.discovered_host_params_filter.filter_params(params, parameter_filter_context, nil)` for the API controller.

lzap replied:
Oh that was it and combination of broken tests. The UI test was in bad shape. Updated, ready for re-review!

lzap commented:
@dLobatog @domcleal guys discovery develop is busted until this is merged: http://ci.theforeman.org/job/test_plugin_matrix/1741/database=postgresql,ruby=2.0.0,slave=fast/console

Thanks!

ohadlevy commented:
what needs to be done? isnt it simply adding back the protected_attributes
gem to discovery?

lzap commented:
> what needs to be done? isnt it simply adding back the protected_attributes
> gem to discovery?

Sorry what? Please follow

https://github.com/theforeman/foreman/pull/3659

for more info.

## 

Later,
 Lukas #lzap Zapletal

dLobatog commented on app/controllers/api/v2/discovered_hosts_controller.rb:
```diff
@@ -93,15 +94,16 @@ def destroy
       param :facts, Hash, :required => true, :desc => N_("hash containing facts for the host with minimum set of facts: discovery_bootif, macaddress_eth0, ipaddress, ipaddress_eth0, interfaces: eth0 (example in case primary interface is named eth0)")
 
       def facts
+        facts = params['facts']
```
Does this change anything?

dLobatog commented:
Looks good to me now, tests pass, PXELess, regular PXE, and auto-provision with rules all worked flawlessly so I'm merging this

dLobatog commented:
Merged as 34db0a489cbd5e8d1cdb5da5ef079828f829fdf1, thanks @lzap!
